src/vs/editor/browser/services/treeSitter/treeSitterParserService.ts

```diff
@@ -0,0 +1,296 @@
+		this._snapshot = undefined;
+	}
+	private _snapshotChunks: { chunk: string; startOffset: number }[] = [];
+	get snapshotChunks() { return this._snapshotChunks; }
+	public addSnapshotChunk(chunk: string, startOffset: number) {
+		this._snapshotChunks.push({ chunk, startOffset });
+	}
+}
+
+export class TreeSitterParserService extends Disposable implements ITreeSitterParserService {
+	readonly _serviceBrand: undefined;
+	private _init: Promise<void>;
+	private _treeSitterTrees: DisposableMap<ITextModel, TreeSitterTree> = new DisposableMap();
+	private _languages: Map<string, Parser.Language> = new Map();
+
+	constructor(@IModelService private readonly _modelService: IModelService,
+		@IFileService private readonly _fileService: IFileService,
+		@ITelemetryService private readonly _telemetryService: ITelemetryService,
+		@ILogService private readonly _logService: ILogService,
+		@IConfigurationService private readonly _configurationService: IConfigurationService
+	) {
+		super();
+
+		const setting = this._getSetting();
+		if (setting.length === 0) {
+			this._init = Promise.resolve();
+			return;
+		}
+
+		this._init = Parser.init({
+			locateFile(_file: string, _folder: string) {
+				const wasmPath: AppResourcePath = `vs/base/common/web-tree-sitter/tree-sitter.wasm`;
+				return FileAccess.asBrowserUri(wasmPath).toString(true);
+			}
+		});
+		// Eventually, this should actually use an extension point to add tree sitter grammars, but for now they are hard coded in core
+		if (setting.includes('html')) {
+			this._addGrammar('html', 'tree-sitter-html');
+		}
+		if (setting.includes('typescript')) {
+			this._addGrammar('typescript', 'tree-sitter-typescript');
+		}
+		this._registerModelServiceListeners();
+
+	}
+
+	private _getSetting(): string[] {
+		return this._configurationService.getValue<string[]>(EDITOR_EXPERIMENTAL_PREFER_TREESITTER) || [];
+	}
+
+	getLanguageLocation(languageId: string): AppResourcePath | undefined {
+		const grammarName = TreeSitterTokenizationRegistry.get(languageId);
+		if (!grammarName) {
+			return undefined;
+		}
+		const languageLocation: AppResourcePath = `vs/base/common/treeSitterLanguages/${grammarName?.name}`;
+		return languageLocation;
+	}
+
+	private _registerModelServiceListeners() {
+		this._register(this._modelService.onModelAdded(model => {
+			this._registerModelListeners(model);
+		}));
+		this._register(this._modelService.onModelRemoved(model => {
+			this._treeSitterTrees.deleteAndDispose(model);
+		}));
+		this._modelService.getModels().forEach(model => this._registerModelListeners(model));
+	}
+
+	private async _registerModelListeners(model: ITextModel) {
+		await this._init;
+		const disposables = new DisposableStore();
+		disposables.add(model.onDidChangeContent(e => this._onDidChangeContent(model, e)));
+		disposables.add(model.onDidChangeLanguage(e => this._onDidChangeLanguage(model, e)));
+		const parser = new Parser();
+		parser.setTimeoutMicros(300 * 1000); // 300 ms
```

Comment:
@hediet is 300 ms a reasonable length of time for the timeout?

Reply:
I think it should be ~20 ms, so that the experience is still somewhat smooth if treesitter is busy for a while

Reply:
I ended up going down to 50ms in the current head of this branch. I'll further decrease it to 20ms.